### PR TITLE
Refactor Build, Versioning into `azure-sdk-tools`

### PR DESCRIPTION
**Pull request by scbedd**
This PR...

- [x] Refactors scripts present under `eng/versioning` to namespace `ci_tools.versioning` within the azure-sdk-tools package. All Integration points in CI are flipped over to consume entrypoints defined in the setup.py.
- [x] Refactors build_packages.py and build_package.py into single module `ci_tools.build`. This function `create_package` within this module is used wherever packages are built. 
- [x] Moves anything related to `setup.py` parsing into `ci_tools.parsing`. More specific, allows users access to `ParsedSetup` which has a classmethod `from_path` that can be aimed at any setup.py. Instead of a crazy tuple that we can get lost in the sauce in, it's just strongly typed members of a class now.
- [x] Moves anything related to processing of requirements into `ci_tools.functions`. 
- [x] Type and doc comment all the moved common functionality

This will be the most painful of the refactoring effort, but we will be able to make iterative fixes quite quickly once this one is in.

Due to the relative risk. I am delaying this merge until post-August-release-week. This is a significant improvement, but nothing that can't wait for a week.

Contributes to #23146

### CODE REVIEW

File: doc/eng_sys_checks.md [hunk 1]
```diff
@@ -32,7 +32,10 @@
 
 There are various tests currently enabled in Azure pipeline for Python SDK and some of them are enabled only for nightly CI checks. We also run some static analysis tool to verify code completeness, security and lint check.
 
-Check the [contributing guide](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md#building-and-testing) for an intro to `tox`.
+Check the [contributing guide](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md#building-and-testing) for an intro to `tox`. For a deeper dive into the tooling that enables the CI checks below and additional detail on reproducing builds locally please refer to the azure-sdk-tools README.md.
+
+TODO: publish github.io
+The development documentation for this repository is published at []().
```
Comment (mccoyp) [inline doc/eng_sys_checks.md:38]: Is this something that can be done now, or needs to be done after getting this merged?

**Review by mccoyp — APPROVED** [review]

File: doc/eng_sys_checks.md (hunk 1 above)
Comment (scbedd) [reply doc/eng_sys_checks.md:38]: I will clean up todos, but this piece will get done after merged. I will merge this today after resolving the regression failures.

**Comment by scbedd** [comment]
The failures present on `python - core` regression tests align with existing failures. Not treating as a blocker.

**Comment by scbedd** [comment]
The single failure during release is due to a known issue with the SemVer powershell scripts, and is _not_ related to the refactor. Overriding and merging.

**Pull request merged**